### ISSUE TITLE
[v2.0.0-beta.1] call beforeEach hook router init

### ISSUE THREAD

**Issue by bartlomieju** [vuejs/vue-router#581]
Hi!

I love this new simple design of vue-router 2.0! My feature request concerns `beforeEach` hook. I would expect it is called on every route including "root routes", eg. `({path: '', component: MyComp}, {path: '/', component: MyComp})`.

My use case is as follows: I have an app that has a static login page (say `xyz.com/login`), after successful login, I am redirected to actual app (`xyz.com/`), and here is where all the good stuff is loaded (vue, vue-router, etc.). I need some info about currently logged in user and that info is crucial to operation of whole app, so I need it before any components are instantiated. So I would create a `beforeEach` hook, in which I check if `this.store.user` is not empty, and if it is I immediately fire a request to API to get that info, and call `next` once this is done. It works as expected if I directly go to some mapped route, but unfortunately not when I go to base roots. 

What are your thoughts on this one?

**Comment by LinusBorg** [comment]
Hi @bartlomieju 

Can you throw together an example on jsfiddle demonstrating the issue? That would be helpful.

**Comment by javisperez** [comment]
I'm having the exact same problem, any news/plans on this?

@bartlomieju did you found any solution to this? i have kinda of the same scenario.

beforeEach hooks are not triggering for path: "/" (root path) and i need it to validate the token.

By the way, i'm using v2.0.0-beta.4

Thank you.

**Comment by bartlomieju** [comment]
@javisperez one hacky solution is like this:
Consider "Basic" from "Examples". Instead of calling `$mount()` at instantiation, we can leverage `created` hook. So start with some function that retrieves token from API

```
function retrieveToken() {
  return this.$http.get('/some/url/).then(data => {
    this.token = data
  })  
}
```

Then define `beforeEach`:

```
router.beforeEach((route, redirect, next) => {
  if (!this.token) {
    retrieveToken().then(() => {
      next()
    })
  }
  else {
    next()
  }
})
```

Then during instatiation of Vue you can do some like this:

```
new Vue({
    data: {
        token: null
    },
    router,
    created() {
        retrieveToken().then(() => {
          this.$mount('#app')
        }
    },
    template: `
    ....
  `
})
```

Not the prettiest solution, calling `beforeEach` on _each_ route including `'/'` would simplify it significantly.

**Comment by javisperez** [comment]
hey @bartlomieju thank you! i did it that way (in the created) and is working great.

I just hope they fix that, as this feels like a nasty solution, but at least is working.

**Comment by bartlomieju** [comment]
Look like 4b52b41d1fdf42c5f01dd8b97677fed6e8783984 solves this issue. @yyx990803 can you confirm it?

**Comment by yyx990803** [comment]
@bartlomieju not sure if it's related. In fact I'm not sure what OP's issue is because I haven't seen a repro so far.

**Comment by bartlomieju** [comment]
I'm sorry I am not able to recreate it properly in [JsFiddle](https://jsfiddle.net/pyz2rsr4/). It's calling hooks in there anyway but has some quirky behaviour. 

In my original problem let's just say I added this: 

```
router.beforeEach((route, redirect, next) => {
  console.log('wait 1500ms')
  setTimeout(() => {
    console.log('call next')
    next()
  }, 1500)
})
```

to "basic" example and I did not call it on `/` root. After testing it today it looks like it was not working in `2.0.0-beta.1` and `2.0.0-beta.2` but works as expected from `2.0.0-beta.3`. I guess it's ok to close now. Sorry for the trouble...
